File: backend/enum-jpa-querydsl-app/src/main/java/com/example/enumapp/common/code/CodeEnums.java
package com.example.enumapp.common.code;

import java.util.Arrays;
import java.util.Objects;

public final class CodeEnums {

    private CodeEnums() {
    }

    public static <E extends Enum<E> & CodeEnum> E fromCode(Class<E> type, String code) {
        if (code == null || code.isBlank()) {
            return null;
        }
        return Arrays.stream(type.getEnumConstants())
                .filter(e -> Objects.equals(e.getCode(), code))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException(
                        "Unknown code '%s' for enum %s".formatted(code, type.getSimpleName())
                ));
    }

    public static String toCode(CodeEnum value) {
        return value == null ? null : value.getCode();
    }
}

File: enum-jpa-querydsl-kotlin-app/src/main/kotlin/com/example/enumapp/common/code/CodeEnums.kt
package com.example.enumapp.common.code

object CodeEnums {

    @JvmStatic
    fun <E> fromCode(type: Class<E>, code: String?): E?
        where E : Enum<E>, E : CodeEnum {
        if (code.isNullOrBlank()) {
            return null
        }
        return type.enumConstants
            .firstOrNull { it.code == code }
            ?: throw IllegalArgumentException(
                "Unknown code '%s' for enum %s".format(code, type.simpleName)
            )
    }

    @JvmStatic
    fun toCode(value: CodeEnum?): String? = value?.code
}
